plugins/apiseeds_lyrics/apiseeds_lyrics.py

```diff
@@ -0,0 +1,122 @@
+from functools import partial
+from urllib.parse import quote, urlencode
+
+from PyQt5 import QtWidgets
+from picard import config, log
+from picard.config import TextOption
+from picard.metadata import register_track_metadata_processor
+from picard.ui.options import register_options_page, OptionsPage
+from picard.webservice import ratecontrol
+
+PLUGIN_NAME = 'Apiseeds Lyrics'
+PLUGIN_AUTHOR = 'Andrea Avallone'
+PLUGIN_DESCRIPTION = 'Fetch lyrics from Apiseeds Lyrics, which provides millions of lyrics from artist all around the world. ' \
+                     'Lyrics provided are for educational purposes and personal use only. Commercial use is not allowed. ' \
+                     'In order to use Apiseeds you need to get a free API key at <em>https://apiseeds.com</em>. ' \
+                     'Want to contribute? Check out the project page at <em>https://github.com/avalloneandrea/apiseeds-lyrics</em>!'''
+PLUGIN_VERSION = '1.0.2'
+PLUGIN_API_VERSIONS = ['2.0.0']
+PLUGIN_LICENSE = 'MIT'
+PLUGIN_LICENSE_URL = 'https://opensource.org/licenses/MIT'
+
+
+class ApiseedsLyricsMetadataProcessor(object):
+
+    apiseeds_host = 'orion.apiseeds.com'
+    apiseeds_port = 443
+    apiseeds_rate_limit = 60 * 1000 / 200
+
+    def __init__(self):
+        super(ApiseedsLyricsMetadataProcessor, self).__init__()
+        ratecontrol.set_minimum_delay((self.apiseeds_host, self.apiseeds_port), self.apiseeds_rate_limit)
+
+    def process_metadata(self, album, metadata, track, release):
+
+        apikey = config.setting['apiseeds_apikey']
+        if not apikey:
+            log.debug('{}: API key is missing, please provide a valid value'.format(PLUGIN_NAME))
+            return
+
+        artist = metadata['artist']
+        if not artist:
+            log.debug('{}: artist is missing, please provide a valid value'.format(PLUGIN_NAME))
+            return
+
+        title = metadata['title']
+        if not title:
+            log.debug('{}: title is missing, please provide a valid value'.format(PLUGIN_NAME))
+            return
+
+        apiseeds_path = '/api/music/lyric/{}/{}'.format(artist, title)
+        apiseeds_params = {'apikey': apikey}
+        album._requests += 1
+        log.debug('{}: GET {}?{}'.format(PLUGIN_NAME, quote(apiseeds_path), urlencode(apiseeds_params)))
+
+        album.tagger.webservice.get(
+            self.apiseeds_host,
+            self.apiseeds_port,
+            apiseeds_path,
+            partial(self.process_response, album, metadata),
+            parse_response_type='json',
+            priority=True,
+            queryargs=apiseeds_params)
+
+    def process_response(self, album, metadata, document, reply, error):
+
+        try:
+            lyrics = document['result']['track']['text']
+            metadata['lyrics'] = lyrics
+            log.debug('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
+
+        except:
```

Comment:
Better check explicitly for `KeyError` that's the only one that should be raised here

Reply:
Added this:
```suggestion
        if error:
            log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))
            album._requests -= 1
            album._finalize_loading(error)
            return
```

and this:
```suggestion
        except KeyError:
```

but Picard crashes when the API is unable to find lyrics. How can I investigate the issue?

Reply:
You probably should change it back to catch all exceptions again, but actually add the stacktrace option. This way you will also see what caused your crashes.

Reply:
I have double checked my code, this is working as expected now:
```
@staticmethod
def process_response(album, metadata, document, reply, error):

    if error:
        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
        album._requests -= 1
        album._finalize_loading(None)
        return

    try:
        lyrics = document['result']['track']['text']
        metadata['lyrics'] = lyrics
        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))

    except KeyError:
        log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))

    finally:
        album._requests -= 1
        album._finalize_loading(None)
```

However I notice that those three lines of code are written twice, and I'm not really a fan of code duplications:
```
log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))
album._requests -= 1
album._finalize_loading(None)
```

Therefore I'm more prone to _"ask forgiveness not permission"_: don't check `error`, let `document` be `None`, and catch the `TypeError` raised by the first access to `document`:
```
@staticmethod
def process_response(album, metadata, document, reply, error):

    try:
        lyrics = document['result']['track']['text']
        metadata['lyrics'] = lyrics
        log.info('{}: lyrics found for track {}'.format(PLUGIN_NAME, metadata['title']))

    except (TypeError, KeyError):
        log.info('{}: lyrics NOT found for track {}'.format(PLUGIN_NAME, metadata['title']))

    finally:
        album._requests -= 1
        album._finalize_loading(None)
```

What do you think about it?

Reply:
I think it should be fine. In case of `error != None` this error will be logged anyway also by the web service module.

It also depends a bit on what we expect this plugin to cause in case of an error. My original suggestion used `album._finalize_loading(error)`. This will cause the entire album loading to fail. Probably with a plugin like this we don't want to be so strict. So I'm totally fine with your solution.